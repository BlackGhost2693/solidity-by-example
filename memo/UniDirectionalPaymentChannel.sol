
pragma solidity ^0.5.11;

import "github.com/OpenZeppelin/openzeppelin-contracts/contracts/math/SafeMath.sol";
import "github.com/OpenZeppelin/openzeppelin-contracts/contracts/cryptography/ECDSA.sol";
import "github.com/OpenZeppelin/openzeppelin-contracts/contracts/utils/ReentrancyGuard.sol";

contract UniDirectionalPaymentChannel is ReentrancyGuard {
    using SafeMath for uint;
    using ECDSA for bytes32;

    address payable public payer;
    address payable public payee;

    // NOTE: contract is closed after the contract calls selfdestruct()
    enum Status {
        Open,
        Closing
    }

    Status public status;
    uint challengePeriod;
    uint finalPayeeBalance;
    uint challengeExpiresAt = 2 ** 256 - 1;

    constructor(address payable _payee, uint _challengePeriod) public payable {
        payer = msg.sender;
        payee = _payee;

        challengePeriod = _challengePeriod;
    }

    function verify(
        bytes memory _signature, address _payer, address _contract, uint _payeeBalance
    )
        public pure returns (bool)
    {
        // NOTE: sign with address of this contract to protect agains replay attack on other contracts
        return keccak256(abi.encodePacked(_contract, _payeeBalance))
            .toEthSignedMessageHash()
            .recover(_signature) == _payer;
    }

    modifier onlyPayerOrPayee() {
        require(msg.sender == payer || msg.sender == payee, "Not authorized");
        _;
    }

    modifier checkSignature(bytes memory _signature, uint _payeeBalance) {
        require(
            verify(_signature, payer, address(this), _payeeBalance),
            "Invalid signature"
        );
        _;
    }

    function startExit(uint _payeeBalance, bytes memory _signature)
        public
        onlyPayerOrPayee
        checkSignature(_signature, _payeeBalance)
        nonReentrant
    {
        require(status == Status.Open, "Channel status must be open");
        require(block.timestamp < challengeExpiresAt, "Expired challenge period");

        status = Status.Closing;
        finalPayeeBalance = _payeeBalance;
        challengeExpiresAt = block.timestamp.add(challengePeriod);
    }

    function challengeExit(uint _payeeBalance, bytes memory _signature)
        public
        onlyPayerOrPayee
        checkSignature(_signature, _payeeBalance)
        nonReentrant
    {
        require(status == Status.Closing, "Channel status must be closing");
        require(block.timestamp < challengeExpiresAt, "Expired challenge period");
        require(
            _payeeBalance > finalPayeeBalance,
            "Payee balance must be greater than current final payee balance"
        );

        finalPayeeBalance = _payeeBalance;
        challengeExpiresAt = block.timestamp.add(challengePeriod);
    }

    // NOTE: griefing (payee does not close) => payer can force close after some time
    function close() public onlyPayerOrPayee nonReentrant {
        require(status == Status.Closing, "Channel status must be closing");
        require(block.timestamp >= challengeExpiresAt, "Challenge period has not expired yet");

        // NOTE: potential DOS by payee
        (bool sent,) = payee.call.value(finalPayeeBalance)("");
        require(sent, "Failed to send Ether");

        // NOTE: avoids replay attack by self destruct
        // NOTE: avoids DOS attack by forcing Ether
        selfdestruct(payer);
    }
}
