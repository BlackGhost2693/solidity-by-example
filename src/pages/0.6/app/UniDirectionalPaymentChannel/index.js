import React from "react"
import Example from "../../../../components/Example"
import html from "./index.html.js"
import { SOL_VERSION } from "../../constants"

export default function (props) {
  return (
    <Example
      title="Uni-Directional Payment Channel"
      version={SOL_VERSION}
      description="An example of uni-directional payment channels in Solidity"
      html={html}
    />
  )
}
