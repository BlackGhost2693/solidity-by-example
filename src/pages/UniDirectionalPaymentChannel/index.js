import React from "react"
import Example from "../../components/Example"
import html from "./index.html.js"

export default function (props) {
  return (
    <Example
      title="Uni-Directional Payment Channel"
      description="An example of uin-directional payment channels in Solidity"
      html={html}
    />
  )
}
