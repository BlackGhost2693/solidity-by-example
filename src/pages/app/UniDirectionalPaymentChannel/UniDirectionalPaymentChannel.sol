pragma solidity ^0.6.0;

import "github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v2.5.0/contracts/math/SafeMath.sol";
import "github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v2.5.0/contracts/cryptography/ECDSA.sol";
import "github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v2.5.0/contracts/utils/ReentrancyGuard.sol";


contract UniDirectionalPaymentChannel is ReentrancyGuard {
    using SafeMath for uint;
    using ECDSA for bytes32;

    address payable public payer;
    address payable public payee;

    uint public expiresAt;

    constructor(address payable _payee, uint _expiresAt) public payable {
        require(_expiresAt > block.timestamp, "Expiration must be > now");

        payer = msg.sender;
        payee = _payee;

        expiresAt = _expiresAt;
    }

    function verify(
        bytes memory _signature,
        address _payer,
        address _contract,
        uint _payeeBalance
    ) public pure returns (bool) {
        // NOTE: sign with address of this contract to protect agains
        // replay attack on other contracts
        return
            keccak256(abi.encodePacked(_contract, _payeeBalance))
                .toEthSignedMessageHash()
                .recover(_signature) == _payer;
    }

    modifier checkSignature(bytes memory _signature, uint _payeeBalance) {
        require(
            verify(_signature, payer, address(this), _payeeBalance),
            "Invalid signature"
        );
        _;
    }

    function close(uint _payeeBalance, bytes memory _signature)
        public
        nonReentrant
        checkSignature(_signature, _payeeBalance)
    {
        require(msg.sender == payee, "Not payee");

        (bool sent, ) = payee.call.value(_payeeBalance)("");
        require(sent, "Failed to send Ether");

        selfdestruct(payer);
    }

    function kill() public {
        require(msg.sender == payer, "Not payer");
        require(block.timestamp >= expiresAt, "channel not expired");
        selfdestruct(payer);
    }
}
