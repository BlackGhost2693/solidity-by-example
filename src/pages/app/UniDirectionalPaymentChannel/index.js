import React from "react"
import Example from "../../../components/Example"
import html from "./index.html.js"

export default function (props) {
  return (
    <Example
      title="Uni-Directional Payment Channel"
      description="An example of uni-directional payment channels in Solidity"
      html={html}
    />
  )
}
