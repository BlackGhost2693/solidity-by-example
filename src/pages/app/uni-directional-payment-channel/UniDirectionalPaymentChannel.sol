// SPDX-License-Identifier: MIT
pragma solidity ^0.8.24;

import "./ECDSA.sol";

contract ReentrancyGuard {
    bool private locked;

    modifier guard() {
        require(!locked);
        locked = true;
        _;
        locked = false;
    }
}

contract UniDirectionalPaymentChannel is ReentrancyGuard {
    using ECDSA for bytes32;

    address payable public sender;
    address payable public receiver;

    uint256 private constant DURATION = 7 * 24 * 60 * 60;
    uint256 public expiresAt;

    constructor(address payable _receiver) payable {
        require(_receiver != address(0), "receiver = zero address");
        sender = payable(msg.sender);
        receiver = _receiver;
        expiresAt = block.timestamp + DURATION;
    }

    function _getHash(uint256 _amount) private view returns (bytes32) {
        // NOTE: sign with address of this contract to protect agains
        // replay attack on other contracts
        return keccak256(abi.encodePacked(address(this), _amount));
    }

    function getHash(uint256 _amount) external view returns (bytes32) {
        return _getHash(_amount);
    }

    function _getEthSignedHash(uint256 _amount)
        private
        view
        returns (bytes32)
    {
        return _getHash(_amount).toEthSignedMessageHash();
    }

    function getEthSignedHash(uint256 _amount)
        external
        view
        returns (bytes32)
    {
        return _getEthSignedHash(_amount);
    }

    function _verify(uint256 _amount, bytes memory _sig)
        private
        view
        returns (bool)
    {
        return _getEthSignedHash(_amount).recover(_sig) == sender;
    }

    function verify(uint256 _amount, bytes memory _sig)
        external
        view
        returns (bool)
    {
        return _verify(_amount, _sig);
    }

    function close(uint256 _amount, bytes memory _sig) external guard {
        require(msg.sender == receiver, "!receiver");
        require(_verify(_amount, _sig), "invalid sig");

        (bool sent,) = receiver.call{value: _amount}("");
        require(sent, "Failed to send Ether");
        selfdestruct(sender);
    }

    function cancel() external {
        require(msg.sender == sender, "!sender");
        require(block.timestamp >= expiresAt, "!expired");
        selfdestruct(sender);
    }
}
