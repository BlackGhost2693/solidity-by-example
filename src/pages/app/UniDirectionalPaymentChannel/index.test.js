import { shallow } from "enzyme"
import React from "react"
import UniDirectionalPaymentChannel from "./index"

test("it renders", () => {
  const component = shallow(<UniDirectionalPaymentChannel />)

  expect(component).toMatchSnapshot()
})
