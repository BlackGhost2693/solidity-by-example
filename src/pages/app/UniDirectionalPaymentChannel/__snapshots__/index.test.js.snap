// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<Example
  description="An example of uni-directional payment channels in Solidity"
  html="<p>Payment channels allow participants to repeatedly transfer Ether off chain.</p>
<p>Here is how this contract is used:</p>
<ul>
<li><code>Alice</code> deploys the contract, funding it with some Ether.</li>
<li><code>Alice</code> authorizes a payment by signing a message (off chain) and sends the signature to <code>Bob</code>.</li>
<li><code>Bob</code> claims his payment by presenting the signed message to the smart contract.</li>
<li>If <code>Bob</code> does not claim his payment, <code>Alice</code> get her Ether back after the contract expires</li>
</ul>
<p>This is called a uni-directional payment channel since the payment can go only in a signle direction from <code>Alice</code> to <code>Bob</code>.</p>
<pre><code class=\\"language-solidity\\">pragma solidity ^0.6.0;

import \\"github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v2.5.0/contracts/math/SafeMath.sol\\";
import \\"github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v2.5.0/contracts/cryptography/ECDSA.sol\\";
import \\"github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v2.5.0/contracts/utils/ReentrancyGuard.sol\\";


contract UniDirectionalPaymentChannel is ReentrancyGuard {
    using SafeMath for uint;
    using ECDSA for bytes32;

    address payable public payer;
    address payable public payee;

    uint public expiresAt;

    constructor(address payable _payee, uint _expiresAt) public payable {
        require(_expiresAt &gt; block.timestamp, \\"Expiration must be &gt; now\\");

        payer = msg.sender;
        payee = _payee;

        expiresAt = _expiresAt;
    }

    function verify(
        bytes memory _signature,
        address _payer,
        address _contract,
        uint _payeeBalance
    ) public pure returns (bool) {
        // NOTE: sign with address of this contract to protect agains
        // replay attack on other contracts
        return
            keccak256(abi.encodePacked(_contract, _payeeBalance))
                .toEthSignedMessageHash()
                .recover(_signature) == _payer;
    }

    modifier checkSignature(bytes memory _signature, uint _payeeBalance) {
        require(
            verify(_signature, payer, address(this), _payeeBalance),
            \\"Invalid signature\\"
        );
        _;
    }

    function close(uint _payeeBalance, bytes memory _signature)
        public
        nonReentrant
        checkSignature(_signature, _payeeBalance)
    {
        require(msg.sender == payee, \\"Not payee\\");

        (bool sent, ) = payee.call.value(_payeeBalance)(\\"\\");
        require(sent, \\"Failed to send Ether\\");

        selfdestruct(payer);
    }

    function kill() public {
        require(msg.sender == payer, \\"Not payer\\");
        require(block.timestamp &gt;= expiresAt, \\"channel not expired\\");
        selfdestruct(payer);
    }
}
</code></pre>
"
  title="Uni-Directional Payment Channel"
/>
`;
